llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -597,11 +597,14 @@ void SCEVUnknown::allUsesReplacedWith(Value *New) {
 ///
 /// Since we do not continue running this routine on expression trees once we
 /// have seen unequal values, there is no need to track them in the cache.
-static int
+static std::optional<int>
 CompareValueComplexity(EquivalenceClasses<const Value *> &EqCacheValue,
                        const LoopInfo *const LI, Value *LV, Value *RV,
                        unsigned Depth) {
-  if (Depth > MaxValueCompareDepth || EqCacheValue.isEquivalent(LV, RV))
+  if (Depth > MaxValueCompareDepth)
+    return std::nullopt;
+
+  if (EqCacheValue.isEquivalent(LV, RV))
```

Comment:
This fixes the bug with inconsistent results, but did you check if the cache actually does anything useful in the average case? In the (admittedly limited) examples I looked at, it was almost pure overhead (i.e., almost no hits, but lots of writes and checks).

Reply:
I do think that we should *also* drop the cache, see https://github.com/llvm/llvm-project/pull/100721#issuecomment-2256157129.

Reply:
Oops, missed that. SGTM.